x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticMultiMatchQueryRewriteInterceptor.java

```diff
@@ -0,0 +1,487 @@
+     * Dense vector models typically produce bounded scores (0-1) while sparse vector models and
+     * non-inference fields produce unbounded scores, causing score range mismatches.
+     */
+    private void detectAndWarnScoreRangeMismatch(MultiFieldInferenceInfo inferenceInfo) {
+        ModelRegistry modelRegistry = modelRegistrySupplier.get();
+        // TODO: validate if we need to check if modelRegistry is null or not
+        if (modelRegistry == null) {
+            // Fallback: warn for any mixed semantic_text + non-inference combination
+            // since we can't determine the exact task types
+            if (inferenceInfo.hasNonInferenceFields() && inferenceInfo.getInferenceFields().isEmpty() == false) {
+                HeaderWarning.addWarning(
+                    "Query spans both semantic_text and non-inference fields. "
+                        + "Dense vector models (TEXT_EMBEDDING) produce bounded scores (0-1) while sparse vector models "
+                        + "(SPARSE_EMBEDDING) and non-inference fields produce unbounded scores, which may cause score "
+                        + "range mismatches and affect result ranking. Consider using Linear or RRF retrievers."
+                );
+            }
+            return;
+        }
+
+        // Check if we have any dense vector models mixed with sparse vector models or non-inference fields
+        boolean hasDenseVectorModel = false;
+        boolean hasSparseVectorModel = false;
+        boolean hasNonInferenceFields = inferenceInfo.hasNonInferenceFields();
+
+        // Collect all inference IDs from all fields
+        Set<String> allInferenceIds = new HashSet<>();
+        for (Map<String, InferenceFieldMetadata> indexFields : inferenceInfo.inferenceFieldsPerIndex().values()) {
+            for (InferenceFieldMetadata fieldMetadata : indexFields.values()) {
+                allInferenceIds.add(fieldMetadata.getSearchInferenceId());
+            }
+        }
+
+        // Check task types for each inference ID
+        for (String inferenceId : allInferenceIds) {
+            try {
+                MinimalServiceSettings settings = modelRegistry.getMinimalServiceSettings(inferenceId);
+                if (settings != null) {
+                    TaskType taskType = settings.taskType();
+                    if (taskType == TaskType.TEXT_EMBEDDING) {
+                        hasDenseVectorModel = true;
+                    } else if (taskType == TaskType.SPARSE_EMBEDDING) {
+                        hasSparseVectorModel = true;
+                    }
+                }
+            } catch (Exception e) {
+                // TODO: validate If we can't get model info, skip this inference ID or throw an error
+            }
+        }
+
+        // Emit warning only if we have dense vector model mixed with sparse vector or non-inference fields
+        if (hasDenseVectorModel && (hasSparseVectorModel || hasNonInferenceFields)) {
```

Comment:
Do we really need this granular level of error handling, or should we warn whenever inference IDs are different?

Reply:
Right, we probably don't need to take BM25 into consideration at all when warning, only inference models.